### PR TITLE
[CSL-1869] Improve block(header) server performance

### PR DESCRIPTION
**Pull request by avieth**
Block deserialization is skipped when sending.
Loading headers performance improved (slightly).

It's a real pain, but I've managed to verify the improvements locally.
Block generator in cardano-sl-1.0 generated a chain of poor quality!
I couldn't figure out how to get the system start time right so that
the test nodes wouldn't immediately jump into recovery mode.

Anyway, since the test node was always in recovery mode, I
noticed that the recovery mode block headers server also suffers
from the block deserialization inefficiency: it will load headers
"up" using `GS.loadHeadersUpWhile`, which in fact loads more than
just the header, it deserializes the block with undo. We can fix that
later, as our relays are probably not going to be in recovery mode (I
hope not!) so it won't affect us.

### CODE REVIEW

**Comment by avieth** [comment]
It's quite difficult to get a good isolated profile of the improvements.

I'm going off of the log timestamps (I did a profiling run but oops, I used `-fprof-auto` so the files are full of noise). With this change, the get blocks request is handled and all blocks sent (2200 empty blocks, 1.4mb total) within a second. Prior to the change it took 4. This is over loopback. The get headers response still takes 4-5 seconds, because of what I explained in the pull request summary: when in recovery mode, the headers are retrieved *by way of the blocks*, whereas the headers retrieval within the get blocks listener goes by way of headers. That extra block decoding in the former is what kills the performance.

**Comment by avieth** [comment]
I should mention as well that I noticed the client took a lot of time to verify the (empty) blocks, perhaps because there was something wrong with them (generator gave me a bad chain). We should think about getting some performance profiles for verification.

I just skimmed the code for this but it looks like verification could also do a lot of block deserialization: `DB.loadBlundsFromTipWhile`.

**Comment by volhovm** [comment]
> it will load headers "up" using GS.loadHeadersUpWhile, which in fact loads more than just the header

It's partially (only one handler is optimized, maybe we can do more) solved here: https://github.com/input-output-hk/cardano-sl/pull/2090

File: node/src/Pos/DB/Block.hs [hunk 1]
```diff
@@ -148,17 +150,17 @@ loadDataWhile
...
     -> m (NewestFirst [] a)
-loadDataWhile getter predicate start = NewestFirst <$> doIt start
+loadDataWhile getter predicate start = NewestFirst <$> doIt [] start
   where
-    doIt :: HeaderHash -> m [a]
-    doIt h
-        | h == genesisHash = pure []
+    doIt :: [a] -> HeaderHash -> m [a]
+    doIt !acc h
+        | h == genesisHash = pure (reverse acc)
```
Comment (volhovm) [inline node/src/Pos/DB/Block.hs]: What happened here? Previously we were doing the right fold, now we're doing left fold and reverse. What for?

File: node/src/Pos/Binary/Communication.hs [hunk 2]
```diff
@@ -63,6 +67,12 @@ instance (HasConfiguration, SscHelpersClass ssc) => Bi (MsgBlock ssc) where
             1 -> MsgNoBlock <$> decode
             t -> fail $ "MsgBlock wrong tag: " <> show t
 
+-- | Get an encoded MsgBlock from an encoded block.
```
Comment (volhovm) [inline node/src/Pos/Binary/Communication.hs]: The comment is incorrect, you're just constructing the prefix here. You could transform it into function `AsBinary Block -> ByteString` or change the comment.

Comment (avieth) [reply node/src/Pos/Binary/Communication.hs]: Right, stale comment.

File: node/src/Pos/DB/Block.hs (hunk 1 above)
Comment (avieth) [reply node/src/Pos/DB/Block.hs]: It performs better.

Comment (volhovm) [reply node/src/Pos/DB/Block.hs]: How? Why do you think so?

Comment (avieth) [reply node/src/Pos/DB/Block.hs]: I profiled it to verify, using it in a pure computation to get 10000 "block headers" (`Int`s). This changed dropped it from ~4ms to ~3ms. I don't know exactly why. I believe the issue was with the `fmap`s through the recursion. What we end up with is `fmap (d0 :) . fmap (d1 :) . fmap (d2 :) . ... . fmap (dn :) (pure [])`. Maybe inlining would help but it depends upon the choice of monad. And anyway, using the accumulator makes this tail recursive, from which I believe GHC can produce faster code.

**Comment by gromakovsky** [comment]
Also please add issue id to your commit message.

**Comment by avieth** [comment]
Updated the commit messages, and set time-warp-nt revision to use cardano-sl-1.0 branch.

I know, pkgs/default.nix must be updated, but whenever I run generate locally it hangs. So my strategy is to wait for travis to fail, and apply the patch it produces for me :)

**Comment by avieth** [comment]
> LGTM

@dcoutts this means nothing without an official GitHub Approval ™

**Comment by jmitchell** [comment]
Initial sanity checks on a new public relay that was intentionally kept out of the RR DNS pool are promising. Michael and Neil were able to send transactions back and forth, Michael starting with no local blockchain and Neil syncing from a relatively recent block.

We have just added that relay to the RR DNS to acquire more subscribers. After the connection count evens out with existing public relay averages we can start comparing metrics. Historically this has taken ~12h, but that was before the topology grew so much; it's only a rough estimate.

**Pull request merged**